### PR TITLE
Add a black box local cluster harness

### PR DESCRIPTION
**Pull request by aeyakovenko**
#### Problem

Multinode tests are not true integration tests.  Changing an implementation detail without actually breaking functionality breaks tests that verify cluster functionality.  The tests are huge, and a painful to maintain.

#### Summary of Changes

Add a harness to write tests against the public interface to the network.

Our code base should contain 2 kinds of tests

* Pure unit tests - These tests cover the implementation details of the function.  If the implementation details change, these tests can be deleted and new ones added if necessary.  There is no backwards compatibility requirement for unit tests.

* Pure integration tests -   They start form a cluster entry point, and a keypair with some funds.  These tests cover network functionality against the public interfaces available on the network.  The tests are written in such a way that they can run against a public testnet or a local cluster.   Just like unit tests, they check 1 thing at a time against the network.

Not every feature needs an integration test.  Especially things that are not fully baked.  But all code needs unit tests.

Fixes #

### CODE REVIEW

**Comment by garious** [comment]
Can you put a hypothetical Greg hat on for a minute with focus on naming and consistentcy? I'll review after your next commit.

File: src/local_cluster.rs [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+        while let Some(node) = self.network_nodes.pop() {
+            node.0.join().expect("join");
+        }
+        for path in &self.ledger_paths {
+            remove_dir_all(path).unwrap();
+        }
+    }
+
+    fn send_tx_and_retry_get_balance(
+        leader: &NodeInfo,
```
Comment (garious) [inline src/local_cluster.rs]: leader_node_info

File: src/local_cluster.rs [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+        alice: &Keypair,
```
Comment (garious) [inline src/local_cluster.rs]: alice_keypair

File: src/local_cluster.rs [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        bob_pubkey: &Pubkey,
+        transfer_amount: u64,
```
Comment (garious) [inline src/local_cluster.rs]: lamports

File: src/local_cluster.rs [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+        expected: Option<u64>,
```
Comment (garious) [inline src/local_cluster.rs]: expected what?

File: src/lib.rs [hunk 5]
```diff
@@ -31,6 +31,7 @@ pub mod blockstream;
 pub mod blockstream_service;
 pub mod blocktree_processor;
 pub mod cluster_info;
+pub mod cluster_tests;
```
Comment (garious) [inline src/lib.rs]: Prefix your additions with`#[cfg(test)]`?

**Comment by garious** [comment]
Biggest naming nit here is the mix of the term "cluster" and "network" in the same module. Looks like every use of "network" can either be "cluster" or deleted.

File: src/local_cluster.rs [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+        self.exit();
+        while let Some(node) = self.network_nodes.pop() {
+            node.0.join().expect("join");
+        }
+        for path in &self.ledger_paths {
+            remove_dir_all(path).unwrap();
+        }
+    }
+
+    fn send_tx_and_retry_get_balance(
```
Comment (CriesofCarrots) [inline src/local_cluster.rs]: send_transaction_and_retry_get_balance

File: src/cluster_tests.rs [hunk 7]
```diff
@@ -0,0 +1,26 @@
+use crate::client::mk_client;
+use crate::contact_info::ContactInfo;
+use crate::gossip_service::converge;
+use solana_sdk::signature::{Keypair, KeypairUtil};
+use solana_sdk::system_transaction::SystemTransaction;
+
+/// Spend and verify from every node in the network
+pub fn spend_and_verify_all_nodes(entry: &ContactInfo, alice: &Keypair, nodes: usize) {
```
[the remark this answers is not in the document]
Comment (aeyakovenko) [reply src/cluster_tests.rs]: @garious the entry point needs to discover the network.  maybe rename `converge` to `discover`?

File: src/local_cluster.rs [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+        let (thread, exit, _) = leader_server.start(None);
+        let mut network_nodes = vec![(thread, exit)];
+        for _ in 0..(num_nodes - 1) {
+            let keypair = Arc::new(Keypair::new());
+            let validator_pubkey = keypair.pubkey();
+            let validator_node = Node::new_localhost_with_pubkey(keypair.pubkey());
+            let ledger_path = tmp_copy_blocktree!(&genesis_ledger_path);
+            ledger_paths.push(ledger_path.clone());
+
+            // Send each validator some tokens to vote
```
[the remark this answers is not in the document]
Comment (aeyakovenko) [reply src/local_cluster.rs]: @garious we need to fund the validators to vote

File: src/local_cluster.rs [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+                node_tokens,
+                None,
+            )
+            .unwrap();
+            info!(
+                "validator {} balance {}",
+                validator_pubkey, validator_balance
+            );
+
+            let voting_keypair = VotingKeypair::new_local(&keypair);
```
[the remark this answers is not in the document]
Comment (aeyakovenko) [reply src/local_cluster.rs]: |                             ^^^^^^^^^^^^^ the trait `solana_sdk::signature::KeypairUtil` is not implemented for `std::sync::Arc<ring::ec::curve25519::ed25519::signing::KeyPair>`

File: src/local_cluster.rs [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+use solana_sdk::pubkey::Pubkey;
+use solana_sdk::signature::{Keypair, KeypairUtil};
+use solana_sdk::system_transaction::SystemTransaction;
+use std::fs::remove_dir_all;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+use std::thread::JoinHandle;
+
+pub struct LocalCluster {
+    pub mint: Keypair,
```
[the remark this answers is not in the document]
Comment (aeyakovenko) [reply src/local_cluster.rs]: @garious funding_keypair?  This keypair is intended for the consumer of the LocalCluster to enter the network

File: src/lib.rs (hunk 5 above)
Comment (aeyakovenko) [reply src/lib.rs]: @garious tests doesn't seem to build that way
```
   Compiling solana v0.12.0 (/home/aeyakovenko/solana)
error[E0432]: unresolved import `solana::cluster_tests`
 --> tests/local_cluster.rs:3:5
  |
3 | use solana::cluster_tests;
  |     ^^^^^^^^^^^^^^^^^^^^^ no `cluster_tests` in the root

error[E0432]: unresolved import `solana::local_cluster`
 --> tests/local_cluster.rs:4:13
  |
4 | use solana::local_cluster::LocalCluster;
  |             ^^^^^^^^^^^^^ could not find `local_cluster` in `solana`

error: aborting due to 2 previous errors
```

Comment (garious) [reply src/lib.rs]: yeah, `local_cluster` too of course

Comment (aeyakovenko) [reply src/lib.rs]: @garious you mean this, right?
```
$ git diff
diff --git a/src/lib.rs b/src/lib.rs
index 7fad53fc..55154a8e 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -31,6 +31,7 @@ pub mod blockstream;
 pub mod blockstream_service;
 pub mod blocktree_processor;
 pub mod cluster_info;
+#[cfg(test)]
 pub mod cluster_tests;
 pub mod db_window;
 pub mod entry;
@@ -43,6 +44,7 @@ pub mod gossip_service;
 pub mod leader_confirmation_service;
 pub mod leader_schedule;
 pub mod leader_schedule_utils;
+#[cfg(test)]
 pub mod local_cluster;
 pub mod local_vote_signer_service;
 pub mod packet;
```

tests depends on lib built without #[cfg(test)] for some reason
```
   Compiling solana v0.12.0 (/home/aeyakovenko/solana)
error[E0432]: unresolved import `solana::cluster_tests`
 --> tests/local_cluster.rs:3:5
  |
3 | use solana::cluster_tests;
  |     ^^^^^^^^^^^^^^^^^^^^^ no `cluster_tests` in the root

error[E0432]: unresolved import `solana::local_cluster`
 --> tests/local_cluster.rs:4:13
  |
4 | use solana::local_cluster::LocalCluster;
  |             ^^^^^^^^^^^^^ could not find `local_cluster` in `solana`

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0432`.
error: Could not compile `solana`.
warning: build failed, waiting for other jobs to finish...
error: build failed
[Finished running. Exit status: 101]
```

File: src/local_cluster.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (garious) [reply src/local_cluster.rs]: Hmm, that's too bad if that's true. A validator should be able to vote with a voting account keypair if it chooses not to delegate its stake. You can leave that part for now. I'll take a look. I've been wanting to delete this copypasta for a long time now.

File: src/local_cluster.rs [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+        for _ in 0..(num_nodes - 1) {
+            let validator_keypair = Arc::new(Keypair::new());
+            let voting_keypair = VotingKeypair::new_local(&validator_keypair);
+            let validator_pubkey = validator_keypair.pubkey();
+            let validator_node = Node::new_localhost_with_pubkey(validator_keypair.pubkey());
+            let ledger_path = tmp_copy_blocktree!(&genesis_ledger_path);
+            ledger_paths.push(ledger_path.clone());
+
+            // Send each validator some tokens to vote
+            let validator_balance = Self::transfer(
```
Comment (pgarg66) [inline src/local_cluster.rs]: I think it's not creating the vote accounts for the validators. Now the balance of vote account is used to calculate node stakes (for leader rotations etc).

Comment (aeyakovenko) [reply src/local_cluster.rs]: @pgarg66 can you show me?

Comment (pgarg66) [reply src/local_cluster.rs]: This function (or something similar) needs to be called here for every validator node

https://github.com/solana-labs/solana/blob/master/fullnode/src/main.rs#L51

Comment (pgarg66) [reply src/local_cluster.rs]: This PR changes the way stakes are calculated https://github.com/solana-labs/solana/pull/2992
Earlier it was using node's account balance. Now it's using staking (voting) account balance.

Comment (aeyakovenko) [reply src/local_cluster.rs]: @pgarg66 cool! can you add it?

Comment (pgarg66) [reply src/local_cluster.rs]: 👍

**Pull request merged**